client/beefy/src/aux_schema.rs

```diff
@@ -0,0 +1,433 @@
+{
+	let version: Option<u32> = load_decode(backend, VERSION_KEY)?;
+
+	match version {
+		None => (),
+		Some(1) =>
+			if let Some(mut state) = load_decode::<_, PersistedState<B>>(backend, WORKER_STATE)? {
+				// Overwrite persisted data with newly provided `min_block_delta`.
+				state.set_min_block_delta(min_block_delta);
+				info!(target: "beefy", "🥩 Loading BEEFY voter state from db: {:?}.", state);
+				return Ok(state)
+			},
+		other =>
+			return Err(ClientError::Backend(format!("Unsupported BEEFY DB version: {:?}", other))),
+	}
+
+	// If no persisted state present, start at the latest session boundary with BEEFY finality.
+	// TODO: we should actually start at pallet-beefy "genesis", but we don't know when that was,
+	// so we walk back the chain up to a session length looking for a starting point.
+	migrate_from_version0(backend, runtime, gossip_engine, finality, min_block_delta).await
+}
+
+async fn migrate_from_version0<B, BE, R>(
+	backend: &BE,
+	runtime: &R,
+	gossip_engine: &mut GossipEngine<B>,
+	finality: &mut Fuse<FinalityNotifications<B>>,
+	min_block_delta: u32,
+) -> ClientResult<PersistedState<B>>
+where
+	B: BlockT,
+	BE: Backend<B>,
+	R: ProvideRuntimeApi<B>,
+	R::Api: BeefyApi<B>,
+{
+	AuxStore::insert_aux(backend, &[(VERSION_KEY, CURRENT_VERSION.encode().as_slice())], &[])?;
+
+	let (active, best_grandpa) = wait_for_runtime_pallet(runtime.clone(), gossip_engine, finality)
+		.await
+		.ok_or_else(|| ClientError::Backend("Gossip engine has terminated.".into()))?;
+
+	debug!(target: "beefy", "🥩 pallet available: header {:?} validator set {:?}", best_grandpa, active);
+
+	if active.id() == GENESIS_AUTHORITY_SET_ID {
+		let start = *best_grandpa.number();
+		info!(
+			target: "beefy",
+			"🥩 Loading BEEFY voter state from genesis on what appears to be first startup. \
+			Starting voting rounds at block {:?}.",
+			start
+		);
+		return Ok(PersistedState::initialize(
+			best_grandpa,
+			Zero::zero(),
+			start,
+			active,
+			min_block_delta,
+		))
+	}
+
+	// Walk back the imported blocks and initialize voter either, at the last block with
+	// a BEEFY justification, or at current session's boundary; voter will resume from there.
+	let blockchain = backend.blockchain();
+	let mut header = best_grandpa.clone();
+	let state = loop {
+		if let Some(true) = blockchain
+			.justifications(header.hash())
+			.ok()
+			.flatten()
+			.map(|justifs| justifs.get(BEEFY_ENGINE_ID).is_some())
+		{
+			info!(
+				target: "beefy",
+				"🥩 Initialize BEEFY voter at last BEEFY finalized block: {:?}.",
+				*header.number()
+			);
+			let mut state = PersistedState::initialize(
+				best_grandpa,
+				*header.number(),
+				*header.number(),
+				active,
+				min_block_delta,
+			);
+			// Mark the round as already finalized.
+			if let Some(round) = state.active_round_mut() {
+				round.conclude(*header.number());
+			}
+			break state
+		}
+
+		if let Some(active) = crate::worker::find_authorities_change::<B>(&header) {
```

Comment:
I was thinking that the justification for the authorities-change-block is necessary in BEEFY (as in GRANDPA) and we've just checked for justification in previous if. So probably this one is optional, or there is a case when it'll work? Sorry if I've messed something in my head :)